RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc

```diff
@@ -0,0 +1,1030 @@
+
+    // auto-vectorized
+    for ( unsigned int ivertex = 0; ivertex < nv; ++ivertex) {
+      const auto mult_resz = track_z - vertices.z_[ivertex];
+      const auto mult_rest = track_t - vertices.t_[ivertex];
+      vertices.ei_cache_[ivertex] = botrack_dz2 * ( mult_resz * mult_resz ) + botrack_dt2 * ( mult_rest * mult_rest );
+    }
+  };
+  
+  auto kernel_add_Z = [ nv, Z_init ] (vertex_t const& vertices) -> double
+    {
+      double ZTemp = Z_init;
+      for (unsigned int ivertex = 0; ivertex < nv; ++ivertex) {	
+	ZTemp += vertices.pk_[ivertex] * vertices.ei_[ivertex];
+      }
+      return ZTemp;
+    };
+
+  auto kernel_calc_normalization = [ beta, nv ] (const unsigned int track_num,
+						  track_t & tks_vec,
+						  vertex_t & y_vec ) {
+    auto tmp_trk_pi = tks_vec.pi_[track_num];
+    auto o_trk_Z_sum = 1./tks_vec.Z_sum_[track_num];
+    //auto o_trk_dz2 = tks_vec.dz2_[track_num];
+    //auto o_trk_dt2 = tks_vec.dt2_[track_num];
+    auto o_trk_err_sum = tks_vec.errsum_[track_num];
+    auto tmp_trk_z = tks_vec.z_[track_num];
+    auto tmp_trk_t = tks_vec.t_[track_num];
+    auto obeta =  -1./beta;
+    
+    // auto-vectorized
+    for (unsigned int k = 0; k < nv; ++k) {
+      // parens are important for numerical stability
+      y_vec.se_[k] +=  tmp_trk_pi*( y_vec.ei_[k] * o_trk_Z_sum );      
+      const auto w = tmp_trk_pi * (y_vec.pk_[k] * y_vec.ei_[k] * o_trk_Z_sum) * o_trk_err_sum; 
+      y_vec.sw_[k]  += w;
+      y_vec.swz_[k] += w * tmp_trk_z;
+      y_vec.swt_[k] += w * tmp_trk_t;
+      y_vec.swE_[k] += w * y_vec.ei_cache_[k]*obeta;
+    }
+  };
+  
+  
+  for (auto ivertex = 0U; ivertex < nv; ++ivertex) {
+    gvertices.se_[ivertex] = 0.0;
+    gvertices.sw_[ivertex] = 0.0;
+    gvertices.swz_[ivertex] = 0.0;
+    gvertices.swt_[ivertex] = 0.0;
+    gvertices.swE_[ivertex] = 0.0;
+  }
+  
+   
+  // loop over tracks
+  for (auto itrack = 0U; itrack < nt; ++itrack) {
+    kernel_calc_exp_arg(itrack, gtracks, gvertices);
+    local_exp_list(gvertices.ei_cache_, gvertices.ei_, nv);
+        
+    gtracks.Z_sum_[itrack] = kernel_add_Z(gvertices);
+    if (edm::isNotFinite(gtracks.Z_sum_[itrack])) gtracks.Z_sum_[itrack] = 0.0;
+    // used in the next major loop to follow
+    sumpi += gtracks.pi_[itrack];
+    
+    if (gtracks.Z_sum_[itrack] > 1.e-100){
+      kernel_calc_normalization(itrack, gtracks, gvertices);
+    }
+  }
+  
+  // now update z and pk
+  auto kernel_calc_zt = [  sumpi, nv, this, useRho0 ] (vertex_t & vertices ) -> double {
+    
+    double delta=0;
+    // does not vectorizes
+    for (unsigned int ivertex = 0; ivertex < nv; ++ ivertex ) {
+      if (vertices.sw_[ivertex] > 0.) {
+	//std::cout << " DA2D_vect sw = " << vertices.sw_[ ivertex ] << ' ' << vertices.swz_[ ivertex ] << ' ' << vertices.swt_[ ivertex ] << std::endl;
+	auto znew = vertices.swz_[ ivertex ] / vertices.sw_[ ivertex ];
+	// prevents from vectorizing if 
+	delta += std::pow( vertices.z_[ ivertex ] - znew, 2 );
```

Comment:
if this really prevents vectorization, perhaps an array of deltas in the vertex_t could be used in this loop to be followed by a simple sum at the end.

Reply:
in case it's the if at the beginning of the loop that prevents vectorization, perhaps the comments could be updated.
Would it vectorize if you hint the more likely value?